File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/glib/views/pst/PSTSquareView.scala
package net.sf.latexdraw.glib.views.pst

import net.sf.latexdraw.glib.models.interfaces.shape.ISquare
import net.sf.latexdraw.glib.models.interfaces.shape.IPoint
import net.sf.latexdraw.glib.models.GLibUtilities
import net.sf.latexdraw.util.LNumber

/**
 * Defines a PSTricks view of the LSquare model.
 * @author Arnaud Blouin
 */
class PSTSquareView(sq:ISquare) extends PSTClassicalView[ISquare](sq){

	update

	override def updateCache(position:IPoint, ppc:Float) {
		if(!GLibUtilities.isValidPoint(position) || ppc<1) return

		emptyCache
		val params  = getPropertiesCode(ppc)
		val tl  	= shape.getTopLeftPoint
		val br  	= shape.getBottomRightPoint
		val x1 		 = tl.getX - position.getX
		val x2 		 = br.getX - position.getX
		val y1 		 = position.getY - tl.getY
		val y2 		 = position.getY - br.getY

		if(shape.isRoundCorner)
			params.append(", framearc=").append(LNumber.getCutNumberFloat(shape.getLineArc)) //$NON-NLS-1$

		val rotation = getRotationHeaderCode(ppc, position)

		if(rotation!=null)
			cache.append(rotation)

		cache.append("\\psframe[")	//$NON-NLS-1$
		cache.append(params)
		cache.append(']').append('(')
		cache.append(LNumber.getCutNumberFloat(x2 / ppc)).append(',')
		cache.append(LNumber.getCutNumberFloat(y1 / ppc)).append(')').append('(')
		cache.append(LNumber.getCutNumberFloat(x1 / ppc)).append(',')
		cache.append(LNumber.getCutNumberFloat(y2 / ppc)).append(')')

		if(rotation!=null)
			cache.append('}')
	}
}

File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/glib/views/pst/PSTSquareView.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2015 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.glib.views.pst;

import net.sf.latexdraw.glib.models.GLibUtilities;
import net.sf.latexdraw.glib.models.interfaces.shape.IPoint;
import net.sf.latexdraw.glib.models.interfaces.shape.ISquare;
import net.sf.latexdraw.util.LNumber;
import org.eclipse.jdt.annotation.NonNull;

/**
 * Defines a PSTricks view of the LSquare model.
 */
class PSTSquareView extends PSTClassicalView<ISquare> {

	protected PSTSquareView(final @NonNull ISquare model) {
		super(model);
	}

	@Override
	public String getCode(IPoint position, float ppc) {
		if(!GLibUtilities.isValidPoint(position) || ppc < 1) return "";

		final StringBuilder params = getPropertiesCode(ppc);
		final IPoint tl = shape.getTopLeftPoint();
		final IPoint br = shape.getBottomRightPoint();
		final double x1 = tl.getX() - position.getX();
		final double x2 = br.getX() - position.getX();
		final double y1 = position.getY() - tl.getY();
		final double y2 = position.getY() - br.getY();
		final StringBuilder code = new StringBuilder();

		if(shape.isRoundCorner()) params.append(", framearc=").append(LNumber.getCutNumberFloat(shape.getLineArc())); //$NON-NLS-1$

		final StringBuilder rotation = getRotationHeaderCode(ppc, position);

		if(rotation != null) code.append(rotation);

		code.append("\\psframe[");    //$NON-NLS-1$
		code.append(params);
		code.append(']').append('(');
		code.append(LNumber.getCutNumberFloat(x2 / ppc)).append(',');
		code.append(LNumber.getCutNumberFloat(y1 / ppc)).append(')').append('(');
		code.append(LNumber.getCutNumberFloat(x1 / ppc)).append(',');
		code.append(LNumber.getCutNumberFloat(y2 / ppc)).append(')');

		if(rotation != null) code.append('}');

		return code.toString();
	}
}
